[OpenAPI] Use invariant culture for TextWriter

martincostello opened the pull request:
# Use invariant culture for TextWriter

Use culture-invariant `TextWriter` implementation for OpenAPI.

## Description

- Ensure OpenAPI documents are written to a culture-invariant `TextWriter` implementation.
- Add/update unit/integration tests to ensure the invariant culture is used to generate the OpenAPI document regardless of the current culture.
- Add support for emitting `exclusiveMinimum`/`exclusiveMaximum` instead of `minimum`/`maximum` where relevant based on the values of `RangeAttribute` properties.

~~I couldn't replicate this issue in .NET 10, I'm guessing Microsoft.OpenApi has had some internal refactoring that means that it's no longer using an overload on `TextWriter` that was hitting this (unless I messed up my test to repro it somehow).~~

- Contributes to #60628.
- Contributes to #61965.

martincostello committed:
[OpenAPI] Use invariant culture for TextWriter

Ensure OpenAPI documents are written to a culture-invariant `TextWriter` implementation.

Contributes to #60628.

martincostello commented on src/OpenApi/sample/Controllers/TestController.cs:
```diff
@@ -32,6 +32,13 @@ public IActionResult PostForm([FromForm] MvcTodo todo)
 
+    [HttpGet]
+    [Route("/getcultureinvariant")]
+    public Ok<CurrentWeather> PostTypedResult()
```
```suggestion
    public Ok<CurrentWeather> GetTypedResult()
```

martincostello committed:
Update TestController.cs

martincostello committed:
Update TestController.cs

desjoerd commented:
I think this change should also be applied for GetDocument, but I am not sure.

martincostello commented:
> GetDocument

Where do you mean exactly?

desjoerd commented:
It's not in the filtered solution, but the dotnet tool/executable which extracts the document during build.

desjoerd commented:
Found it: https://github.com/dotnet/aspnetcore/blob/faf269645b0338a9f4a7ecc88b46053f80f6f602/src/Tools/GetDocumentInsider/src/Commands/GetDocumentCommandWorker.cs

martincostello committed:
Apply fix to OpenAPI document tool

Apply the same fix to ensure invariant formatting to the OpenAPI document tool.

Co-Authored-By: Sjoerd van der Meer <2460430+desjoerd@users.noreply.github.com>

captainsafia commented:
> I couldn't replicate this issue in .NET 10, I'm guessing Microsoft.OpenApi has had some internal refactoring that means that it's no longer using an overload on TextWriter that was hitting this (unless I messed up my test to repro it somehow).

Do you mean that the new tests pass without any of the code changes in this PR?

It might be good to track the change on the Microsoft.OpenApi side that caused this.

martincostello committed:
Repro issue

Change the integration test to use HTTP client so that the middleware to write the OpenAPI document is invoked, which is where the bug is.

Going through `OpenApiDocument.SerializeAsJsonAsync()` uses a different code path that already uses a culture-invariant `TextWriter` so wasn't susceptible to the bug.

martincostello committed:
[OpenAPI] Fix Culture formatting for [Range]

Update OpenAPI range formatting to format using the target culture and update tests to write in the invariant culture.

Co-Authored-By: Sjoerd van der Meer <2460430+desjoerd@users.noreply.github.com>

martincostello commented:
@captainsafia I worked out why the repo wasn't happening and fixed it.

TL;DR - because the snapshots directly got the document and self-serialized it, it didn't go through the code path for the endpoint that was causing the issue (the tests _didn't_ have the same bug in them).

captainsafia commented on src/OpenApi/test/Microsoft.AspNetCore.OpenApi.Tests/Integration/snapshots/OpenApi3_0/OpenApiDocumentIntegrationTests.VerifyOpenApiDocument_documentName=controllers.verified.txt:
```diff
@@ -1,9 +1,14 @@
 ﻿{
-  "openapi": "3.0.4",
```
Feels like something spooky happened with the new test data setup. These should still target OpenApi v3.

martincostello replied:
Oh yes, I see what I've done.

There's no mechanism to pass the OpenAPI version to use on to the endpoint now that it's doing it over HTTP instead of using the service directly.

martincostello replied:
Right, I've fixed this now, but I'm not particularly happy with it.

Unless I've missed something, it's not possible to dynamically determine the OpenAPI version to return without encoding it into the document's name, so I've had to expose the endpoints on multiple URLs each to cover 3.0 and 3.1, which then means I need to customise `ShouldInclude` so the paths still get returned.

Is it worth adding a capability somehow to allow the user to dynamically compute the OpenAPI version to use at runtime? I tried to do it dynamically with `IPostConfigureOptions<T>`, but as it uses `IOptionsMonitor<T>` it only runs once and keeps the first request's preference.

Maybe something like this?

```diff
public OpenApiOptions
{
+   public Func<HttpContext, OpenApiSpecVersion> OpenApiVersionSelector { get; set; } = (_) => OpenApiVersion;
}
```

If you think it's worth proposing I'll spin up an API Suggestion issue (and am happy to take a hand at implementing it if approved), assuming there's still time to get it into 10.0.

martincostello replied:
Example: https://github.com/martincostello/aspnetcore/commit/65f941f0b745d6f097f7801479beea84e4750bc8

martincostello committed:
Fix snapshot tests

Make OpenAPI sample endpoints version-specific.

martincostello committed:
Refactor method

Use `Convert.ToString()` instead of a format string.

martincostello committed:
Support exclusive(Minimum|Maximum)

- Add support for emitting `exclusive(Minimum|Maximum)` when `RangeAttribute.(Minimum|Maximum)IsExclusive` is set.
- Add tests for `[Range]` attribute handling.

martincostello committed:
Avoid string parsing

Do not round trip the values from `[Range]` if they were set using `RangeAttribute(int, int)`.

martincostello committed:
Remove unused local

Leftover from refactoring.

martincostello committed:
Add test for invalid ranges

Add test to ensure no exception is thrown if a range contains invalid string values.

martincostello commented:
While working on a related Swashbuckle change I realised that if `RangeAttribute` set `MaximumIsExclusive` or `MinimumIsExclusive` to `true`, the OpenAPI schema wasn't being emitted in a way that respected that.

I've fixed that and added unit tests for various scenarios. I've also refactored the code to avoid the round-trip string format/parse where possible.

martincostello commented:
@captainsafia Could I get a review on this soon please so that this makes it in for preview 7?

martincostello committed:
Revert submodule

Rebase fail.

martincostello committed:
Merge branch 'main' into gh-60628

martincostello committed:
Merge branch 'main' into gh-60628

martincostello committed:
Merge branch 'main' into gh-60628

martincostello committed:
Merge branch 'main' into gh-60628

martincostello committed:
Add dedicated localization tests

Move the fixture for OpenAPI localization to its own fixture to not pollute the sample app.

martincostello commented:
@captainsafia I've moved the localization stuff out into its own text fixture as discussed offline. I've left the changes to extend the sample to have different OpenAPI versions covered in the snapshots.

Merging with main has uncovered a failing test added by #61399 so I need to investigate what's going on there. It fails in main too, so maybe it's another localisation-related thing that my en-GB machine has flushed out.

```text
 OpenApiSchemaServiceTests.RespectsEnumDefaultValueInControllerFormParameters
   Source: OpenApiSchemaService.RequestBodySchemas.cs line 715
   Duration: 11 ms

  Message: 
System.InvalidCastException : Unable to cast object of type 'Status' to type 'System.String'.

  Stack Trace: 
JsonSerializer.UnboxOnWrite[T](Object value)
JsonTypeInfo`1.SerializeAsObject(Utf8JsonWriter writer, Object rootValue)
JsonSerializer.WriteNodeAsObject(Object value, JsonTypeInfo jsonTypeInfo)
JsonNodeSchemaExtensions.ApplyDefaultValue(JsonNode schema, Object defaultValue, JsonTypeInfo jsonTypeInfo) line 166
JsonNodeSchemaExtensions.ApplyParameterInfo(JsonNode schema, ApiParameterDescription parameterDescription, JsonTypeInfo jsonTypeInfo) line 316
OpenApiSchemaService.GetOrCreateUnresolvedSchemaAsync(OpenApiDocument document, Type type, IServiceProvider scopedServiceProvider, IOpenApiSchemaTransformer[] schemaTransformers, ApiParameterDescription parameterDescription, CancellationToken cancellationToken) line 128
OpenApiSchemaService.GetOrCreateSchemaAsync(OpenApiDocument document, Type type, IServiceProvider scopedServiceProvider, IOpenApiSchemaTransformer[] schemaTransformers, ApiParameterDescription parameterDescription, CancellationToken cancellationToken) line 141
OpenApiDocumentService.GetFormRequestBody(OpenApiDocument document, IList`1 supportedRequestFormats, IEnumerable`1 formParameters, IList`1 endpointMetadata, IServiceProvider scopedServiceProvider, IOpenApiSchemaTransformer[] schemaTransformers, CancellationToken cancellationToken) line 567
OpenApiDocumentService.GetRequestBodyAsync(OpenApiDocument document, ApiDescription description, IServiceProvider scopedServiceProvider, IOpenApiSchemaTransformer[] schemaTransformers, CancellationToken cancellationToken) line 516
OpenApiDocumentService.GetOperationAsync(ApiDescription description, OpenApiDocument document, IServiceProvider scopedServiceProvider, IOpenApiSchemaTransformer[] schemaTransformers, CancellationToken cancellationToken) line 312
OpenApiDocumentService.GetOperationsAsync(IGrouping`2 descriptions, OpenApiDocument document, IServiceProvider scopedServiceProvider, IOpenApiOperationTransformer[] operationTransformers, IOpenApiSchemaTransformer[] schemaTransformers, CancellationToken cancellationToken) line 269
OpenApiDocumentService.GetOpenApiPathsAsync(OpenApiDocument document, IServiceProvider scopedServiceProvider, IOpenApiOperationTransformer[] operationTransformers, IOpenApiSchemaTransformer[] schemaTransformers, CancellationToken cancellationToken) line 253
OpenApiDocumentService.GetOpenApiDocumentAsync(IServiceProvider scopedServiceProvider, HttpRequest httpRequest, CancellationToken cancellationToken) line 75
OpenApiDocumentServiceTestBase.VerifyOpenApiDocument(ActionDescriptor action, Action`1 verifyOpenApiDocument, CancellationToken cancellationToken) line 46
OpenApiSchemaServiceTests.RespectsEnumDefaultValueInControllerFormParameters() line 721
--- End of stack trace from previous location ---
```

martincostello commented:
> I need to investigate what's going on there

Worked it out - MVC isn't part of the OpenAPI solution filter, so when I rebased it hadn't been rebuilt with the fix, but the _tests_ were being rebuilt because they are in the filter.

Running `build.cmd` in a terminal to refresh everything fixed it.

martincostello committed:
Merge branch 'main' into gh-60628

captainsafia reviewed (commented):
Bleh, had these comments in pending and never hit submit 😭

martincostello committed:
Address feedback

- Switch integration tests back to using service directly.
- Use dedicated resource for localization test.

martincostello committed:
Apply suggestions from code review

Restore `sealed` as there's no longer a derived class.

martincostello committed:
Revert sample changes

Remove version-specific endpoints.

captainsafia reviewed (approved):
Thanks for reacting to all the feedback here! 😃
